### PR TITLE
FakeTensor cache SymInt support

### PR DESCRIPTION
**Pull request by aorenste**
Adds support for SymInts in the FakeTensor cache.

A couple notes:
1. When a SymInt is present in the input key for a FakeTensor operation we cache on the ShapeEnv instead of using the FakeTensorMode cache. This is necessary so we don't have to remember and check the guards. It reduces the cache hits but there's diminishing return on how much work we can do before the cache becomes more of a burden than a gain.
2. We need to be careful that when we cache an output SymInt that is a direct copy from the input that when we have a cache-hit we copy the SymNode from the input to the output. This is important because the fx-graph building code actually uses SymNode ids in the process of building the graph so constructing a same-content-but-different-id SymNode will fail.
3. In the cache key we store SymInts as a _PySymInputStub. These represent SymInt (and friends) but support `__hash__` and `__eq__` (which SymInt do not).
4. In the cache entry we store SymInts as a _SymIntOutputStub.

Perf example:
```
python benchmarks/dynamo/timm_models.py --ci --accuracy --timing
--explain --inductor --dynamic-shapes --dynamic-batch-only --device cuda
--training --amp --total-partitions 2 --partition-id 0 --output
/tmp/training_timm_models.csv --filter crossvit_9_240
```
fake tensor cache before:
```
INFO: FakeTensor cache stats:
INFO:   cache_hits: 68137
INFO:   cache_misses: 837
INFO:   cache_bypasses:
INFO:     symbolic shape:            48224
INFO:     CompositeImplicitAutograd: 917
INFO:     non-fake tensor:           70
INFO:     non-FakeTensor output:     62
INFO:     non-builtin:               8
INFO:     dynamic output shape:      1
```
and after:
```
INFO: FakeTensor cache stats:
INFO:   cache_hits: 88187
INFO:   cache_misses: 14233
INFO:   cache_bypasses:
INFO:     CompositeImplicitAutograd: 1037
INFO:     non-FakeTensor output:     602
INFO:     non-fake tensor:           70
INFO:     unsafe view:               36
INFO:     non-builtin:               8
INFO:     dynamic output shape:      1
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #131052
* __->__ #127596
* #129780
* #131014

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -86,6 +90,9 @@ class _Unassigned:
 
 _UNASSIGNED = _Unassigned()
 
+_py_sym_type = (SymInt, SymFloat, SymBool)
+_PySymType = Union[SymInt, SymFloat, SymBool]
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:94]: dupe of `py_sym_types` in torch/fx/proxy_tensor - can we deduplicate ?

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -931,15 +932,17 @@ class TensorMetadata:
...
+        for v in self.shape:
+            state.convert_input(result, v)
+        for v in self.stride:
+            state.convert_input(result, v)
         result.append(self.device)
         result.append(self.layout)
         result.append(self.memory_format)
-        result.append(self.storage_offset)
-        result.append(self.storage_bytes)
+        state.convert_input(result, self.storage_offset)
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:944]: so i guess the idea is we use `convert_input` for anything that could be py sym type ? add comment /docstring to flatten_into?

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -956,63 +959,271 @@ def extract_tensor_metadata(t: Tensor) -> TensorMetadata:
...
     memory_format: Optional[torch.memory_format] = suggest_memory_format(t)
-    if is_sparse_any(t) or not t.is_contiguous(memory_format=memory_format):
+    # Don't call is_contiguous() on a Tensor which has symbolic sizes or things
+    # will go badly (guards will be messed up?)
+    if (
+        t._has_symbolic_sizes_strides
+        or is_sparse_any(t)
+        or not t.is_contiguous(memory_format=memory_format)
+    ):
         memory_format = None
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:969]: Should we just remove memory_format from the key then ? Pytorch eager has a property on the tensor to distinguish ambiguous cases I believe and for propagating views of channels last, but it looks like in the symint case we just compute from strides.

File: torch/_subclasses/fake_tensor.py [hunk 4]
```diff
@@ -956,63 +959,271 @@ def extract_tensor_metadata(t: Tensor) -> TensorMetadata:
...
 
     storage_offset = t.storage_offset()
 
     return TensorMetadata(
-        dtype=t.dtype,
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:965]: is this the kwargs thing lol

File: torch/_subclasses/fake_tensor.py [hunk 5]
```diff
@@ -956,63 +959,271 @@ def extract_tensor_metadata(t: Tensor) -> TensorMetadata:
...
 
-    def __hash__(self) -> int:  # type: ignore[override]
+    def __eq__(self, other: object) -> bool:
+        return isinstance(other, _DispatchCacheKey) and self.key == other.key
+
+    def __hash__(self) -> int:
         return self.hashvalue
 
+    def strip_shape_env(self) -> None:
+        # We need to strip the ShapeEnv from any values before we store in the
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:1017]: would it make more sense to use weakrefs ?

File: torch/_subclasses/fake_tensor.py [hunk 6]
```diff
@@ -956,63 +959,271 @@ def extract_tensor_metadata(t: Tensor) -> TensorMetadata:
...
+        t.is_sparse,
+        t.is_coalesced() if t.is_sparse else None,
+        t.dense_dim() if t.is_sparse else None,
+        t.sparse_dim() if t.is_sparse else None,
     )
 
 
-class _DispatchCacheKey(list):
+@dataclass
+class _DispatchCacheKey:
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:996]: Fake Tensor is already quite large... would it make sense to factor this out to `fake_cache_utils` or something along those lines?

File: torch/_subclasses/fake_tensor.py [hunk 7]
```diff
@@ -956,63 +959,271 @@ def extract_tensor_metadata(t: Tensor) -> TensorMetadata:
...
+        )
+
+    def __str__(self) -> str:
+        return str(self._expr)
+
+    def __repr__(self) -> str:
+        return f"_DeconstructedSymNode{{{self._expr!r}, {self.pytype!r}, {self._hint!r}, {self.constant!r}, {self.fx_node!r}}}"
+
+    def __eq__(self, other: object) -> bool:
+        raise NotImplementedError
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:1057]: Why is it we define `_value_eq` but not `__eq__`

File: torch/_subclasses/fake_tensor.py [hunk 8]
```diff
@@ -1030,12 +1243,65 @@ class DispatchCacheInfo:
     Information about the state of the FakeTensor dispatch cache.
     """
 
+    __slots__ = ("hits", "misses", "bypasses", "size")
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:1246]: Annoying we cant juse use py 3.10 `@dataclass(slots=True)`.. in a few years i guess

File: torch/_subclasses/fake_tensor.py [hunk 9]
```diff
@@ -1030,12 +1243,65 @@ class DispatchCacheInfo:
...
+
+    __slots__ = ("sym_node_lookup", "shape_env")
+
+    # We track the SymNodes so when we get the output we can see if it exactly
+    # matches one of the inputs so we can uncache it properly.
+    sym_node_lookup: Dict[int, int]  # id(SymNode) -> index
+
+    # This is the ShapeEnv for one of the cached SymNodes. There is no guarantee
+    # about which node's ShapeEnv it will be.
+    shape_env: Optional[ShapeEnv]
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:1268]: Sorry, would you say more about this `There is no guarantee
    # about which node's ShapeEnv it will be.`

File: torch/_subclasses/fake_tensor.py [hunk 10]
```diff
@@ -1030,12 +1243,65 @@ class DispatchCacheInfo:
...
+
+    def __init__(self, shape_env: Optional[ShapeEnv] = None) -> None:
+        self.sym_node_lookup = {}
+        self.shape_env = shape_env
+
+    def cache_on_shape_env(self) -> bool:
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:1274]: hmm, still dont really follow this

**Review by eellison — COMMENTED** [review]
Looks good! Couple questions:

- Did you think about using weakrefs and would simplify some of the logic ?

- was there any observable compilation speedup ?

File: torch/_subclasses/fake_tensor.py (hunk 7 above)
Comment (aorenste) [reply torch/_subclasses/fake_tensor.py:1057]: It has to have the same protocol as SymNode.  I'll add a comment.

File: torch/_subclasses/fake_tensor.py (hunk 4 above)
Comment (aorenste) [reply torch/_subclasses/fake_tensor.py:965]: yup

File: torch/_prims_common/__init__.py [hunk 11]
```diff
@@ -1835,7 +1835,7 @@ def are_strides_like_channels_last(
     for d in dim_order:
         if guard_size_oblivious(shape[d] == 0):
             return False
-        if strides[d] < min:
+        if guard_size_oblivious(strides[d] < min):
```
[the remark this answers is not in the document]
Comment (aorenste) [reply torch/_prims_common/__init__.py:1838]: nothing smart - tests failed, lol.

**Comment by aorenste** [comment]
> Did you think about using weakrefs and would simplify some of the logic ?

I did.  We can't put it on the SymNode -> ShapeEnv ref because that would be a change to SymNode.  We can't put it on the cache itself because then dropping the SymNode would drop it from the cache, losing the cache benefit.

> was there any observable compilation speedup ?

I measured a previous version it on the HUD and (assuming I was reading it correct) it was a 7% speedup on the cudagraphs_dynamic tests.

File: torch/_subclasses/fake_tensor.py (hunk 5 above)
Comment (aorenste) [reply torch/_subclasses/fake_tensor.py:991]: It doesn't really give us an advantage. We'd still need a non-SymInt structure to hold the value since SymInt has to have a non-weak ShapeEnv so we'd need a version with and without the weakref.

File: torch/_subclasses/fake_tensor.py (hunk 3 above)
Comment (aorenste) [reply torch/_subclasses/fake_tensor.py:944]: Trying it in #131052

File: torch/_subclasses/fake_tensor.py [hunk 12]
```diff
@@ -1245,8 +1283,9 @@ def _cached_dispatch_impl(
             else:
                 self._validate_cache_key(func, args, kwargs)
                 output = self._dispatch_impl(func, types, args, kwargs)
-                entry = self._make_cache_entry(key, func, args, kwargs, output)
-                FakeTensorMode.cache[key] = entry
+                entry = self._make_cache_entry(state, key, func, args, kwargs, output)
+                key.strip_shape_env()
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:1287]: would be great if we did not explicitly have to invoke `strip_shape_env` so that users of this could not get into a bad state. maybe either the key could do it in its `__init__` or this could be handled as in the dict as part of the `__setitem__` call

**Review by eellison — APPROVED** [review]
🚢

File: torch/_subclasses/fake_tensor.py (hunk 12 above)
Comment (aorenste) [reply torch/_subclasses/fake_tensor.py:1287]: I get the concern - but there's only one place where we need to call `strip_shape_env()` and that's just before we insert the value into the cache... so the danger area is pretty contained.

I hesitate to use `__setitem__` just because that adds a bunch of complexity/potential slowdown.

File: torch/fx/experimental/sym_node.py [hunk 13]
```diff
@@ -123,7 +123,7 @@ def compute_hint():
                 "Cannot create SymNode of type "
                 f"{pytype} with incompatible hint of type {type(hint)}"
             )
-            if self.shape_env._translation_validation_enabled:
+            if self.shape_env and self.shape_env._translation_validation_enabled:
```
Comment (ezyang) [inline torch/fx/experimental/sym_node.py:126]: Why can shape_env be None here now?

Comment (aorenste) [reply torch/fx/experimental/sym_node.py:126]: I think that was debug cruft - I'll remove in a follow-up PR.

File: torch/fx/experimental/sym_node.py [hunk 14]
```diff
@@ -138,15 +138,30 @@ def compute_hint():
...
         )
 
+    def _value_eq(self, other: "SymNode") -> bool:
+        # Purposely don't include the shape_env in the eq.
+        return (
+            self._expr == other._expr
+            and self.pytype == other.pytype
+            and self._hint == other._hint
+            and self.constant == other.constant
+            and self.fx_node == other.fx_node
```
Comment (ezyang) [inline torch/fx/experimental/sym_node.py:158]: I actually think you probably only need expr and pytype, the rest are derived quantities that don't matter for the purpose of cache matching. Actually, pytype also probably not needed either since you're probably only ever looking at integral quantities.

File: torch/_prims_common/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (aorenste) [reply torch/_prims_common/__init__.py:1838]: test/dynamo/test_dynamic_shapes.py::DynamicShapesMiscTests::test_unbacked_auto_functionalize_op_dynamic_shapes

File: torch/fx/experimental/sym_node.py (hunk 14 above)
Comment (aorenste) [reply torch/fx/experimental/sym_node.py:158]: If we were going to do that I'd want to move this check over to fake_tensor.py instead of having it on SymNode. It seems kind of weird to have a function on SymNode that was specific to caching fake tensors (and not have it named something like `value_eq_for_fake_tensor_cache`)

Comment (eellison) [reply torch/fx/experimental/sym_node.py:158]: I think i had to use pytype for symbool previously

File: torch/_subclasses/fake_tensor.py [hunk 15]
```diff
@@ -1478,40 +1534,65 @@ def _output_from_cache_entry(
...
+        if isinstance(storage_bytes, SymInt):
+            # Do it this way so we don't import symbolic_shapes (which imports
+            # expensive sympy) unless we have to.
+            from torch.fx.experimental.symbolic_shapes import guard_size_oblivious
+
+            zero_bytes = guard_size_oblivious(storage_bytes == 0)
+        else:
+            zero_bytes = storage_bytes == 0
+        if zero_bytes:
             empty.untyped_storage().resize_(0)
```
Comment (ezyang) [inline torch/_subclasses/fake_tensor.py:1596]: What is going on here? I'm more asking about the preexisting code. Why do we need a special case for storage bytes zero?
